substrate/executor/src/wasm_executor.rs

```diff
@@ -60,23 +64,45 @@ impl Heap {
 }
 
 struct FunctionExecutor<'e, E: Externalities + 'e> {
+	sandbox_store: sandbox::Store,
 	heap: Heap,
 	memory: MemoryRef,
+	table: Option<TableRef>,
 	ext: &'e mut E,
 	hash_lookup: HashMap<Vec<u8>, Vec<u8>>,
 }
 
 impl<'e, E: Externalities> FunctionExecutor<'e, E> {
-	fn new(m: MemoryRef, e: &'e mut E) -> Result<Self> {
+	fn new(m: MemoryRef, t: Option<TableRef>, e: &'e mut E) -> Result<Self> {
 		Ok(FunctionExecutor {
+			sandbox_store: sandbox::Store::new(),
 			heap: Heap::new(&m)?,
 			memory: m,
+			table: t,
 			ext: e,
 			hash_lookup: HashMap::new(),
 		})
 	}
 }
 
+impl<'e, E: Externalities> sandbox::SandboxCapabilities for FunctionExecutor<'e, E> {
+	fn store(&self) -> &sandbox::Store {
+		&self.sandbox_store
+	}
+	fn allocate(&mut self, len: u32) -> u32 {
+		self.heap.allocate(len)
+	}
+	fn deallocate(&mut self, ptr: u32) {
+		self.heap.deallocate(ptr)
+	}
+	fn write_memory(&mut self, ptr: u32, data: &[u8]) -> ::std::result::Result<(), DummyUserError> {
+		self.memory.set(ptr, data).map_err(|_| DummyUserError)
+	}
+	fn read_memory(&self, ptr: u32, len: u32) -> ::std::result::Result<Vec<u8>, DummyUserError> {
+		self.memory.get(ptr, len as usize).map_err(|_| DummyUserError)
```

Comment:
It's supposed that the sandboxed function doesn't have access to a supervisor's (aka contract-runtime) memory.

Reply:
but it seems that they are using the same memory instance?

Reply:
Hm, this function is called only by supervisor 

Take a look at the place where this function is called
https://github.com/paritytech/polkadot/pull/140/files#diff-ab114f06bdcdf15b97c22509b94564a9R272

It's called within `GuestExternals::invoke_index` method. This method is invoked when guest function calls an imported functon. But the `GuestExternals::invoke_index` itself is executed in host context (wasm-executor) and reads memory from supervisor only for getting the result of the execution supervisor function in order to it to guest module.

The guest module doesn't have a control over a pointer that is read from supervisor memory (since this pointer comes from the supervisor itself) and thus can't read arbitrary data from the supervisor's memory.